internal/server/helpers_test.go

```diff
@@ -0,0 +1,274 @@
+		IssuedFor:  admin.ID,
+		ProviderID: data.InfraProvider(tx).ID,
+		ExpiresAt:  time.Now().Add(1000 * time.Second),
+	}
+
+	accessKey, err := data.CreateAccessKey(tx, token)
+	assert.NilError(t, err)
+
+	assert.NilError(t, tx.Commit())
+	return organizationData{
+		Organization:   otherOrg,
+		Admin:          admin,
+		AdminAccessKey: accessKey,
+	}
+}
+
+func adminAccessKey(s *Server) string {
+	for _, id := range s.options.Users {
+		if id.Name == "admin@example.com" {
+			return id.AccessKey
+		}
+	}
+
+	return ""
+}
+
+// withAdminUser may be used with setupServer to setup the server
+// with an admin identity and access key
+func withAdminUser(_ *testing.T, opts *Options) {
+	opts.Users = append(opts.Users, User{
+		Name:      "admin@example.com",
+		AccessKey: "BlgpvURSGF.NdcemBdzxLTGIcjPXwPoZNrb",
+	})
+	opts.Grants = append(opts.Grants, Grant{
+		User:     "admin@example.com",
+		Role:     "admin",
+		Resource: "infra",
+	})
+}
+
+func withSupportAdminGrant(_ *testing.T, opts *Options) {
+	opts.Grants = append(opts.Grants, Grant{
+		User:     "admin@example.com",
+		Role:     "support-admin",
+		Resource: "infra",
+	})
+}
+
+func withMultiOrgEnabled(_ *testing.T, opts *Options) {
+	opts.DefaultOrganizationDomain = "example.com"
+	opts.EnableSignup = true
+}
+
+func createAdmin(t *testing.T, db data.WriteTxn) *models.Identity {
+	user := &models.Identity{
+		Name: "admin+" + generate.MathRandom(10, generate.CharsetAlphaNumeric),
+	}
+	err := data.CreateIdentity(db, user)
+	assert.NilError(t, err)
+
+	err = data.CreateGrant(db, &models.Grant{
+		Subject:   uid.NewIdentityPolymorphicID(user.ID),
+		Resource:  "infra",
+		Privilege: models.InfraAdminRole,
+	})
+	assert.NilError(t, err)
+
+	return user
+}
+
+func createAccessKey(t *testing.T, db data.GormTxn, email string) (string, *models.Identity) {
+	t.Helper()
+	user := &models.Identity{Name: email}
+	err := data.CreateIdentity(db, user)
+	assert.NilError(t, err)
+
+	provider := data.InfraProvider(db)
+
+	token := &models.AccessKey{
+		IssuedFor:  user.ID,
+		ProviderID: provider.ID,
+		ExpiresAt:  time.Now().Add(10 * time.Second),
+	}
+
+	body, err := data.CreateAccessKey(db, token)
+	assert.NilError(t, err)
+
+	return body, user
+}
+
+func createIdentities(t *testing.T, db data.GormTxn, identities ...*models.Identity) {
+	t.Helper()
+	for i := range identities {
+		err := data.CreateIdentity(db, identities[i])
+		assert.NilError(t, err, identities[i].Name)
+		for _, g := range identities[i].Groups {
+			err := data.AddUsersToGroup(db, g.ID, []uid.ID{identities[i].ID})
+			assert.NilError(t, err)
+		}
+		assert.NilError(t, err, identities[i].Name)
+	}
+}
+
+func createGroups(t *testing.T, db data.GormTxn, groups ...*models.Group) {
+	t.Helper()
+	for i := range groups {
+		err := data.CreateGroup(db, groups[i])
+		assert.NilError(t, err, groups[i].Name)
+	}
+}
+
+func createUser(t *testing.T, srv *Server, routes Routes, email string) *api.CreateUserResponse {
```

Comment:
It might be out of scope but what's the difference between `createUser` and `createIdentities`? It looks like `createUser` creates a user with API while `createIdentities` creates a user with data. Something to consider since these look very similar on the surface

Reply:
Ya, there is some cleanup to do here.

I think `createIdentities` is a good pattern. It's a convenient way of creating a user with data that is relevant for the test case.

`createUser` is different from most since it uses the API to create things, and only lets you set the email address.